pkg/resource/plugin/provider_plugin.go

```diff
@@ -257,10 +257,14 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
 		return DiffResult{}, err
 	}
 
-	// If this function is called, we must have complete configuration for the underlying provider. Per DiffConfig,
-	// any unknown input will cause the provider to be replaced, which will cause all of its resources to be replaced,
-	// and we do not call `Diff` for resources that are being replaced due to a change to their provider reference.
-	contract.Assert(p.cfgknown)
+	// If the configuration for this provider was not fully known--e.g. if we are doing a preview and some input
+	// property was sourced from another resource's output properties--don't call into the underlying provider.
+	// Instead, indicate that the diff is unavailable and write a message
+	if !p.cfgknown {
+		logging.V(7).Infof("%s: cannot diff due to unknown config", label)
+		err := errors.New("cannot diff: this resource's provider has unknown configuration values")
+		return DiffResult{Changes: DiffUnavailable}, err
```

Comment:
I don't think I like returning both an error and a non-empty `DiffResult` here - would it be sufficient to simply return `DiffUnavailable` here (or even something like `DiffUnavailableUnknownConfig`) and emit a warning in the step generator if you see it? It looks to me the error here is only for an error message and we might as well produce the warning text in the step generator.

Reply:
I do want the error text to be generated by the provider, as the provider has the best knowledge of why the diff cannot be computed. For example, we could go so far as to list the unknown config keys in the message.

I agree that what is in here isn't pretty. The other approach I considered was to use a distinguished error type for this, but this was expedient. If that approach sounds better, I can refactor.

Reply:
I'm fine with what you've got, I guess. I prefer distinguished error types in general because of Go's utter lack of any kind of sum type (which would be my favorite solution for `DiffUnavailable`)...

Reply:
Done.